### ISSUE TITLE
MongoDB ArrayFilters support (MongoDB 3.6.1)

### ISSUE THREAD

**Issue by jbreed** [Automattic/mongoose#5986]
I'm having issues implementing ArrayFilters using Mongoose. Looking at the dependencies the listed MongoDB version is 3.0.5. I have 3.6.1 installed on the server along with replacing all node_module files for mongodb with the latest incase there was something not updating.

ArrayFilters appears to have been implemented in version 3.5.12+ and was unsure if my hangup is with Mongoose, or my syntax.
https://jira.mongodb.org/browse/SERVER-831

Example of syntax used:
```
var conditions = {"_id": r.id}
var update= {
    $set: {"this.$[i].that.$[x].value" : newValue}
}
var options = {
    arrayFilters: [{"i._id": thisId}, {"x.Id": thatId}],
    multi:true
}
myModel.update(conditions,update, options, function(error, rowsAffected){
    // handler
}
```

Node.js version: v8.9.1
NPM version: 5.5.1
MongoDB version: 3.6.1
mongoose@5.0.0-rc2
    - appears to use mongodb 3.0.1

**Comment by jbreed** [comment]
When doing the line from the Mongo shell it is successful with static entries for _id variables when I enclose them inside of 'ObjectId'. As Node is stating the object ID's are strings I tried the mongoose.Types.ObjectId(thisId), but have the same result. I am getting a result of 0,0,0

**Comment by jbreed** [comment]
At this time I simply moved over to using .save instead. I would prefer to use update() and will migrate back to this if I can find a solution that works.

**Comment by actraiser** [comment]
I replied to your stackoverflow post but for the sake of completeness, I copy and paste my suggestion here as well:

I don't know if Mongoose 5.0.0 is supposed to support Arrayfilters out of the box but you can achieve it by using Mongoose's command-method which directly executes on MongoDB, hence can utilize all features available which includes ArrayFilters on MongoDB 3.6.1

**Example:**

    mongoose.connection.db.command({
      update: <YourModel>.collection.name,
      updates: [
        {
          q: { 'field1.field2._id': mongoose.Types.ObjectId(<someObjectid>) },
          u: {
            $set: { 'field1.$.field2.$[field].fieldToUpdate': "updated!" },
          },
          arrayFilters: [
            { 'field._id': mongoose.Types.ObjectId(<someObjectid>) },
          ],
        },
      ],
    })

**Comment by jbreed** [comment]
Thanks @actraiser! I switched over to .save but I'll give this a go as id prefer to use the update method.

**Comment by vkarpov15** [comment]
This fix will be in 5.0.0, will backport for 4.13.10

**Comment by jbreed** [comment]
@vkarpov15 Thanks! I haven't switch from .save yet but before putting into production I'll be coming back to using update

**Comment by maustand** [comment]
Hi @vkarpov15 
Just i would like to know if still this feature is not supported ?
Still it is on development ?

I am using mongoose v4.13.9 and basicly it ignores my arrayFilters..

Thank you so munch !

**Comment by vkarpov15** [comment]
@maustand correct haven't merged a release for it yet, will be in 4.13.10. Thanks for your patience, we've been focused on finishing up the 5.0.0 release.

**Comment by vkarpov15** [comment]
So upon further inspection we won't be able to support this properly in the 4.x branch because [v2.x of the node driver does not support MongoDB 3.6](https://docs.mongodb.com/ecosystem/drivers/driver-compatibility-reference/#node-js-driver-compatibility). Even if mongoose 4.x were to support array filters, it depends on v2.x of the mongodb driver, which doesn't send down `arrayFilters` to the mongodb server, so you'd still get this error even if mongoose 4.x did the right thing:

```
MongoError: No array filter found for identifier 'element' in path 'grades.$[element]'
```

TLDR; unfortunately, you're going to have to upgrade to mongoose 5.0.0 to use array filters with mongodb 3.6.

**Comment by pellea** [comment]
Using mongoose 5.0.10 I still have this issue. 
The query seems fine but the data is not modified: `{ n: 1, nModified: 0, ok: 1 }`.

@vkarpov15 What's the status of this?

**Comment by Bexanderthebex** [comment]
Hello @vkarpov15, I encountered this issue in v5.3.2 using mongodb 4.0 

here is the following code I used:
`    await User.updateOne(
      { _id: userId, 'courses.course_id': courseId },
      {
        $set: {
          'courses.$[].course_progress.$[cp].status': 'IN_PROGRESS',
        },
      },
      {
        arrayFilter: [
          {
            'cp.chapter_order_no': 1,
            'cp.content_order_no': 1,
          },
        ],
      }
    );`

> "No array filter found for identifier 'cp' in path 'courses.$[].course_progress.$[cp].status'

**Comment by lineus** [comment]
@Bexanderthebex it looks like you're passing in an option called `arrayFilter` to your query, I believe this should be `arrayFilters`.

**Comment by Bexanderthebex** [comment]
Still the same error after changing arrayFilter to _arrayFilters_ tho

**Comment by vkarpov15** [comment]
Can you show your schemas please? Also, why does 'cp.content_order_no' appear twice?

**Comment by Bexanderthebex** [comment]
Hi @vkarpov15, the other one is 'cp.chapter_order_no'

`` courses: {
    type: [
      {
        course_id: {
          type: mongoose.Schema.Types.ObjectId,
          required: true,
        },
        organization_id: {
          type: mongoose.Schema.Types.ObjectId,
          required: true,
        },
        status: {
          type: {
            enum: ['DONE', 'IN_PROGRESS', 'TO_TAKE'],
          },
          required: true,
        },
        // understanding is, if there is no expiry field, then course has no expiry
        expiry: {
          type: Date,
        },
        last_course_content: {
          // course schema - chapters.order
          chapter_order_no: {
            type: Number,
            min: [1, 'starting chapter order number should be 1'],
            default: 1,
            required: true,
          },
          // course schema - chapters.contents.content_order
          content_order_no: {
            type: Number,
            min: [1, 'starting content order number should be 1'],
            default: 1,
            required: true,
          },
        },
        complete_in_order: {
          type: Boolean,
          required: true,
        },
        course_progress: {
          type: [
            {
              content_id: {
                type: mongoose.Schema.Types.ObjectId,
                required: true,
              },
              content_title: {
                type: String,
                required: true,
              },
              duration: {
                type: Number,
                required: true,
              },
              version: {
                type: Number,
                required: true,
              },
              content_static_storage_address: {
                type: String,
                required: true,
              },
              status: {
                type: {
                  enum: ['DONE', 'IN_PROGRESS', 'TO_TAKE'],
                },
                default: 'TO_TAKE',
              },
              chapter_order_no: {
                type: Number,
                required: true,
              },
              content_order_no: {
                type: Number,
                required: true,
              },
              video_progress: { type: Number },
              last_viewed: { type: Date },
            },
          ],
        },
      },
    ],
  },
``

**Comment by lineus** [comment]
@Bexanderthebex The error in the image you shared says that it's looking for an identifier called `sample` not `cp`. Is there a place where you have sample instead of cp in your code?

I created [this working gist](https://gist.github.com/lineus/1945ca7f142bb9c3ca4f251e42c1b86a) based on the relevant parts of your example/schema.

**Comment by Bexanderthebex** [comment]
@lineus that is because I tried doing running this instead:

`    await User.update(
      { _id: userId, 'courses.course_id': courseId },
      {
        $set: {
          'courses.$[sample].course_progress.$[cp].status': 'IN_PROGRESS',
        },
      },
      {
        arrayFilters: [
          {
            'cp.chapter_order_no': 1,
            'cp.content_order_no': 1,
          },
        ],
      }
    );`

Isn't this that it could demonstrate that mongoose can't recognize any of the filters?

**Comment by Bexanderthebex** [comment]
I will try your gist

**Comment by Bexanderthebex** [comment]
@lineus can we you make a working gist where you will filter the object in the 'top' array?

**Comment by lineus** [comment]
@Bexanderthebex I think I see what's going on, in your query.... yeah that :)

**Comment by lineus** [comment]
@Bexanderthebex [Here is a working gist](https://gist.github.com/lineus/9b17b14031b47938d2050f0fe0bbc43f) that moves the `top` _id condition from the query to a second arrayFilter.

**Comment by Bexanderthebex** [comment]
Ahh that did it. Would want an explanation why this worked:

`    await User.update(
      { _id: userId, 'courses.course_id': courseId },
      {
        $set: {
          'courses.$.status': 'IN_PROGRESS',
        },
        $inc: {
          'courses.$.last_course_content.chapter_order_no': 1,
          'courses.$.last_course_content.content_order_no': 1,
        },
      }
    );`

**Comment by lineus** [comment]
`$[]` means all elements in the array. `$` is the first matching element from the query doc.

**Comment by Bexanderthebex** [comment]
Got it!
